src/Language/PureScript/Environment.hs

```diff
@@ -129,10 +130,23 @@ makeTypeClassData
   -> [FunctionalDependency]
   -> Bool
   -> TypeClassData
-makeTypeClassData args m s deps = TypeClassData args m s deps determinedArgs coveringSets
+makeTypeClassData args m s deps = TypeClassData args m' s deps determinedArgs coveringSets
   where
     ( determinedArgs, coveringSets ) = computeCoveringSets (length args) deps
 
+    coveringSets' = S.toList coveringSets
+
+    m' = map (\(a, b) -> (a, b, addVtaInfo b)) m
+    
+    addVtaInfo :: SourceType -> [[Text]]
+    addVtaInfo memberTy = do
+      let mentionedArgIndexes = S.fromList (mapMaybe argToIndex $ freeTypeVariables memberTy)
+      let leftovers = map (`S.difference` mentionedArgIndexes) coveringSets'
+      map (map (fst . (args !!)) . S.toList) leftovers
```

Comment:
It probably returns `[[], []]` in some contexts.

> If it were `Maybe (Set (NonEmpty Text))`, that would represent at the type level that there are no duplicates or empty variables-needed lists, and then my original confusion about what the Nothing case adds wouldn't apply.

Does the order of the `NonEmpty Text` affect whether two non empty lists are the same? I presume yes?

But otherwise, yeah, this would work and I don't need to define an ADT for this.

Reply:
> Does the order of the `NonEmpty Text` affect whether two non empty lists are the same? I presume yes?

Yes, but that shouldn't matter if you always generate these by filtering down the same list of class arguments. You'll never get the same non-empty list of arguments in a different order. (And it's good that you're keeping them in a list and not an [`NESet`](https://hackage.haskell.org/package/nonempty-containers-0.3.4.4/docs/Data-Set-NonEmpty.html) or something, so that they can be presented in the order in which their definition refers to them and not alphabetical order.)

Reply:
When I converted things to just `Set (NonEmpty Text)`, converting the `Set` back to a list produces an order that differs from the one in which they were inserted. The `ClassHeadNoVTA5.purs` revealed that change. The two ty vars are `tyNotAppearInBody` and `norThisOne`, but when I make this change, the test output prints `norThisOne` first rather than `tyNotAppearInBody`. 

I could add the index of the arg (i.e. `Set (Int, NonEmpty Text)`), but then this stops feeling `Set`y due to the additional metadata. Using `Map (NonEmpty Text) Int` seems to be the correct way to account for all of this at the type-level: uniqueness of `NonEmpty Text` value, `Int` metadata stored, difference in meaning between empty and non-empty. However, it seems a bit misleading since we don't use it to look up anything, just track additional metadata about the arg index so we can sort it properly later when converting it back to a List...

Thoughts?

Reply:
> is there a difference between `Nothing` and `Just Set.empty`?

Yeah, `Set.empty` is an impossible result, unless you use it as an unnatural sentinel. The set returned represents the distinct ways to clarify an ambiguous method call; if that set is empty, that naturally means that the method call is impossible to disambiguate, which doesn't make sense.

So yeah you *could* drop the `Maybe` and use `Set.empty` in place of `Nothing`, but IMO using the `Nothing` to indicate that the method doesn't need disambiguation is clearer.

> When I converted things to just `Set (NonEmpty Text)`, converting the `Set` back to a list produces an order that differs from the one in which they were inserted.

Well, yeah; does that order matter? If so, I think I'd use `Maybe (Set (NonEmpty Int))`, and add the args as `[Text]` to the `UnknownsWithVtaRequiringArgs` constructor, doing the lookup when the error is presented.

Reply:
> Well, yeah; does that order matter? If so, I think I'd use Maybe (Set (NonEmpty Int)), and add the args as [Text] to the UnknownsWithVtaRequiringArgs constructor, doing the lookup when the error is presented.

The order doesn't matter per say, but it breaks my expectation. If I define a class
```purs
class MultiFdBidi tyNotAppearInBody norThisOne | tyNotAppearInBody -> norThisOne, norThisOne -> tyNotAppearInBody where
  useMultiFdBidi :: Int
```
and the error I get for this situation lists `norThisOne` first instead of `tyNotAppearInBody`, it surprises me because it doesn't match the order of the fun deps in the type class. 

The information presented doesn't change; I still have to clarify one of those to select a type class. However, I start to wonder whether there's a bug in the compiler because the order differs.

>> is there a difference between Nothing and Just Set.empty?
>
> Yeah, `Set.empty` is an impossible result, unless you use it as an unnatural sentinel

Right, it's impossible when this value is understood in its context. But without that context, it can still represent an impossible state via `Just Set.empty`. So, I'm asking why we're allowing such an impossible state to be represented if we use `Maybe (Set x)` rather than `Maybe (NESet x)`. And since I've already raised the order issue above, I guess I'm leaning towards `Maybe (NonEmptyMap (NonEmpty Text) Int)` (but likely providing a `newtype NESetWithMetadata k v = NESetWithMetadata (NonEmptyMap k v)` to clarify that the map isn't meant for lookup) so as to represent a value with these characteristics:
- `Nothing` = no ambiguity
- `Just` = ambiguous
    - keys represent unique values to prevent duplication
    - values represent additional metadata, so we can display them in the original order of fun deps when converting the set of keys back to a list

This might be overengineering this small problem, but all documentation for it seems stored within the type itself.

Reply:
I guess the only argument I can think of against my idea is the cost of implementing it:
- an added dependency on `nonemptyMap` for this single usage when an alternative is 90% good enough
- a new module (that may have name clashes with others in the ecosystem) for `NESetWithMetadata` so it can be reused in more places since it's not specific to the compiler per say, but which doesn't have a complete API because we don't need that additional API here.

Welp. `Maybe (Map (NonEmpty Text) Int)` it is I guess.

Reply:
> If so, I think I'd use Maybe (Set (NonEmpty Int)), and add the args as [Text] to the UnknownsWithVtaRequiringArgs constructor, doing the lookup when the error is presented.

Rereading this comment. So, rather than storing the text of the args as I'm currently doing, this would store the index to the corresponding args in the set. In other words, storing the metadata itself rather than the identifier of the type variable. When constructing the error, we lookup what the identifier is using the index. This bypasses the need to thread that additional metadata around at the cost of some unsafety that should be impossible (i.e. the lookup returning a `Maybe`).

Reply:
So... I thought the output was based on the order of the fun deps. If it was, then the above suggestion would still produce a wrong output order because the above suggestion only looks at the type class' head's type variables. But having updated `ClassHeadNoVTA6.purs` to show a better example of multiple sets and changing the fun deps order to highlight this and updating the code to account for that, I saw that the output didn't change. Then I realized it was not dependent on the order of the fun deps but just the args themselves.

So, yes, you're suggestion is the better approach. And sorry for not understanding it earlier.